tools/eslint-plugin/dist/rules/no-internal-barrel-imports.js

```diff
@@ -0,0 +1,297 @@
+  },
+
+  create(context) {
+    const parserServices = getParserServices(context);
+    /** @type {ts.Program} */
+    const program = parserServices.program;
+    const checker = program.getTypeChecker();
+    const maybeTsConfig = program.getCompilerOptions().configFilePath;
+
+    const extraOpts = context.options[0];
+    const ignoredBarrelModules = (
+      (maybeTsConfig &&
+        extraOpts &&
+        extraOpts[OPTION_IGNORED_BARREL_MODULES]) ||
+      []
+    ).map((p) => {
+      /** @type {string} earlier check means it is definitely a string */
+      const tsConfig = maybeTsConfig;
+      return OsPaths.normalize(OsPaths.resolve(OsPaths.dirname(tsConfig), p));
+    });
+
+    return {
+      ImportDeclaration(node) {
+        /** @param {ts.Symbol | undefined} symbol */
+        function getRelativeImportPathForExportedSymbol(symbol) {
+          if (symbol === undefined) return "";
+          const declaration = symbol.valueDeclaration || symbol.declarations[0];
+          const fileOfExport = declaration.getSourceFile();
+          const path = TsImportPaths.normalize(
+            withoutExt(
+              TsImportPaths.relative(
+                TsImportPaths.dirname(thisModule.fileName),
+                fileOfExport.fileName
+              )
+            )
+          );
+          // a path unprefixed by "./" is interpretted as from node_modules
+          const ensurePrefixed = (path) =>
+            path[0] === "." ? path : `./${path}`;
+          return ensurePrefixed(path);
+        }
+
+        if (typeof node.source.value !== "string")
+          throw Error("Invalid input source");
+
+        /** @type{ts.ImportDeclaration} */
+        const importNodeTs = parserServices.esTreeNodeToTSNodeMap.get(node);
+        if (!importNodeTs)
+          throw Error("equivalent typescript node could not be found");
+
+        if (importNodeTs.importClause && importNodeTs.importClause.isTypeOnly)
+          return;
+
+        const thisModule = importNodeTs.getSourceFile();
+
+        if (!thisModule.resolvedModules.has(importNodeTs.moduleSpecifier.text))
+          throw Error("module did not have this import");
+
+        const importInfo = thisModule.resolvedModules.get(
+          importNodeTs.moduleSpecifier.text
+        );
+
+        const importIsPackage =
+          importInfo === undefined || importInfo.isExternalLibraryImport;
+        if (importIsPackage) return;
+
+        // path case can be different between the cased import specifier, and the resolved system path
+        // use lower case on OSs known to use lower cased file systems (e.g. windows and mac but not linux)
+        const targetFilePathOfThisImportStmt =
+          process.platform === "win32" || process.platform === "darwin"
+            ? importInfo.resolvedFileName.toLowerCase()
+            : importInfo.resolvedFileName;
+
+        const importedModule = program.getSourceFileByPath(
+          targetFilePathOfThisImportStmt
+        );
```

Comment:
have you looked if there's anything in the TypeChecker class that could help you with this? If you could resolve the import statement, you could then check if the target is an export statement or a declaration/etc.
I guess you could also get the parent SourceFile object and check if it has any other export statements in it, but I'm not sure if that's necessary. Just checking if you're importing things directly or through a re-export should be enough

Reply:
I don't currently see a more ergonomic way to do this, I've been looking around, are you against us keeping it this way?

Reply:
I'd say if you really can't find a way to use TypeChecker here, leave it this way. There's nothing inherently wrong with it, but in my experience using Typescript's built in functions is always much faster than trying to parse the AST yourself. Could you time how long it takes to run this rule in a big package, such as imodeljs-common or imodeljs-frontend? You can do that by setting the TIMING environment variable and running eslint:
https://eslint.org/docs/1.0.0/developer-guide/working-with-rules#per-rule-performance
if it appears in top 10 of the rules we currently have enabled, that's a sign you should think about optimizing it.

Reply:
it's not in the top 10

Reply:
if we ever want to increase the speed of our lint step, we should look at that

Reply:
I'm not even sure why we have react rules in `imodeljs-recommended`, they should probably be only in the `ui` config. @calebmshafer perhaps you remember why we put them here?

Reply:
also I'd like some clarification on this comment, I'm not sure what you're referring to as ast parsing in this module. I did add one instance of AST single-depth traversal (`ts.forEachChild`) to remove reliance on the `.imports` internal property after you wrote this comment.

Reply:
and we pretty much always want to increase the speed of it because it's painfully slow with some of the 3rd party rules we're using... but as always it boils down to how much development effort it requires and how much it actually improves. Since you already tried and couldn't find anything helpful in TypeChecker, I don't think it's worth spending any more time on this unless it gets really high up on the list of slowest rules.

Reply:
I won't resolve this for a bit, in case we use it as a forum to discuss the inclusion of react rules in the lint step for core/backend

Reply:
regarding the react rules, I think most projects shouldn't use the rule configurations we're recommending for react... so I'd be in favor of just removing them, but we could maybe offer both imodeljs-recommended and imodeljs-react-recommended eslint configs

Reply:
I mean basically anything that involves iterating through multiple "candidates", be it `importedModule.imports` or `ts.forEachChild`. Most TypeChecker operations are as simple as a hash table look up because the compiler resolves most references in one go, so you don't have to iterate every time an import declaration is found. Basically if you use the compiler's built-in resolving and if I understand it correctly, it should only run the type checker once (which is required and shared by other rules, so it won't really add any time to your rule) and then the complexity of your rule is O(n), whereas iterating through imports makes it more like O(n*m). It's not a tremendous difference because m is probably infinitesimal compared to n, but it could still make your rule twice as slow.

Reply:
I see what you mean now, well right now it doesn't seem to be a bottleneck at all, but if we ever see it pop up in the slow rules we can look at refining it.

Reply:
I'm going to resolve this, we can use somewhere else more relevant to discuss the unnecessary inclusion of react linting rules in imodeljs-backend